Improve serialization of tuples and sets

bolkedebruin opened the pull request:
Sets cannot be encoded into JSON directly
and tuples loose information when serialized into JSON directly. We now serialize both into a dict and encode them properly.

@uranusjr 

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

ashb commented on airflow/serialization/serde.py:
```diff
@@ -260,11 +262,11 @@ def _stringify(classname: str, version: int, value: T | None) -> str:
-    elif isinstance(value, _iterables):
-        s += ",".join(str(serialize(value, False)))
+    elif isinstance(value, _collections):
+        s += ",".join(str(serialize(value)))
```
`str()` + `join` here looks suspect.  (since `",".join("abc") == "a,b,c"` )

ashb replied:
A nice addition to the tests along side the round trip tests would be to verify what the serialized form ends up as

bolkedebruin replied:
Good catch. Tests added

bolkedebruin committed:
Improve serialization of tuples and sets

Sets cannot be encoded into JSON directly
and tuples loose information when serialized into JSON directly.
We now serialize both into a dict and encode them properly.

uranusjr commented on airflow/serialization/serializers/builtin.py:
```diff
@@ -0,0 +1,59 @@
+    if version > __version__:
+        raise TypeError("serialized version is newer than class version")
+
+    if classname == qualname(tuple):
+        return tuple(data)
+
+    if classname == qualname(set):
+        return set(data)
+
+    if classname == qualname(frozenset):
+        return frozenset(data)
```
```python
for t in stringifiers:
    if qualname(t) == classname:
        return t(data)
```

Does this make sense?

bolkedebruin replied:
The intention is not to arbitrarily instantiate a class and only allow `[stringifiers]`. I'm not entirely happy yet with the approach, so if you have a better idea I am all ears :-)

uranusjr commented on airflow/serialization/serde.py:
```diff
@@ -211,6 +211,9 @@ def deserialize(o: T | None, full=True, type_hint: Any = None) -> object:
     if CLASSNAME in o and VERSION in o:
         classname, version, value = decode(o)
 
+    if classname == "":
```
```suggestion
    if not classname:
```

Or is esactly an empty string somehow significant?

bolkedebruin replied:
I think mypy was borking here but don't quite remember, maybe my GoLang influences came in here too much :-). I'll try your suggestion

uranusjr commented on airflow/serialization/serde.py:
```diff
@@ -266,10 +269,13 @@ def _stringify(classname: str, version: int, value: T | None) -> str:
+        print("COLL")
         s += ",".join(deserialize(value, full=False))
     elif isinstance(value, dict):
+        print(f"DICT {value}")
```
Accidentally left in?

bolkedebruin replied:
argh, I cannot say no now, can I? :-).

bolkedebruin committed:
Improve stringify, add tests

bolkedebruin committed:
Checking

bolkedebruin committed:
add comment

bolkedebruin committed:
remove debug stuff

bolkedebruin committed:
Fix dataset typo

uranusjr committed:
Keep linters happy

uranusjr committed:
Keep linters happy

uranusjr commented on airflow/serialization/serde.py:
```diff
@@ -207,19 +216,22 @@ def deserialize(o: T | None, full=True, type_hint: Any = None) -> object:
 
     if CLASSNAME in o and VERSION in o:
```
Should this be an `elif`? It seems weird we try to still infer the classname ane version from `o` twice when `type_hint == True`.

bolkedebruin replied:
No, as a provided serialized classname overrides the type_hint. It really is a hint, we don't force the conversion. It is provided here for when we move the old serialization (DAGs) to the new serialization. The old serialization doesn't store type information for the top-level object (e.g. a serialized dag) so the code allows us to have:

```
s_dag = serialize(dag, type_hint=SerializedDAG)
```

and have proper versioning working and keep backwards compatibility. But both `classname` and `version` need to be present in order for this to be true. The new format stipulates the use of 'VERSION' hence we know what we are dealing with.

uranusjr committed:
Some more style linter issues

uranusjr committed:
Raise not return

uranusjr committed:
Fix test

bolkedebruin committed:
Add limitations

bolkedebruin committed:
Little bit of style
